tools/lib/tests/test_logreader.py

```diff
@@ -216,6 +218,43 @@ def test_auto_source_scenarios(self, is_internal):
       log_len = len(list(lr))
       self.assertEqual(qlog_len, log_len)
 
+  def test_sort_by_time(self):
+    msgs = list(LogReader(f"{TEST_ROUTE}/0/q"))
+    self.assertNotEqual(msgs, sorted(msgs, key=lambda m: m.logMonoTime))
+
+    msgs = list(LogReader(f"{TEST_ROUTE}/0/q", sort_by_time=True))
+    self.assertEqual(msgs, sorted(msgs, key=lambda m: m.logMonoTime))
+
+  def test_only_union_types(self):
+    with tempfile.NamedTemporaryFile() as qlog:
+      # write valid Event messages
+      num_msgs = 100
+      msgs = [log.Event.new_message() for _ in range(num_msgs)]
+      with open(qlog.name, "wb") as f:
+        f.write(b"".join(msg.to_bytes() for msg in msgs))
+
+      msgs = list(LogReader(qlog.name))
+      self.assertEqual(len(msgs), num_msgs)
+      [m.which() for m in msgs]
+
+      # append non-union Event message
+      event_msg = log.Event.new_message()
+      non_union_bytes = bytearray(event_msg.to_bytes())
+      non_union_bytes[event_msg.total_size.word_count * 8] = 0xff  # set discriminant value out of range using Event word offset
+      with open(qlog.name, "ab") as f:
+        f.write(non_union_bytes)
+
+      # ensure new message is added, but is not a union type
+      msgs = list(LogReader(qlog.name))
+      self.assertEqual(len(msgs), num_msgs + 1)
+      with self.assertRaises(capnp.KjException):
+        [m.which() for m in msgs]
+
+      # should not be added when only_union_types=True
+      msgs = list(LogReader(qlog.name, only_union_types=True))
+      self.assertEqual(len(msgs), num_msgs)
+      [m.which() for m in msgs]
+
```

Comment:
You can't initialize an invalid union type, so this is the only way I could figure out how to do it. It's huge and kinda fragile, but I think it's worth it since we use this in the pipeline.